### DISCUSSION TITLE
Nautilus CFD Subscriber

### DISCUSSION THREAD

**Discussion by skalaydzhiyski** [nautechsystems/nautilus_trader#1438] [General]
Hi guys, I am pretty new to nautilus trader, but I have managed to set it up with interactive brokers with STKs and it works like a charm. Hands down the best back testing platform I have tried + the ability to trade the same strategies live is just great.

Anyway all is good with Stocks and with the examples on the repo, but for Interactive Brokers I mainly need to be able to trade CFD contracts, but there is literally nothing I managed to find online.

Can someone please show me a very tiny working example of a "subscriber" kind of a strategy running live in nautilus_trader where it prints the current price of NVDA for example ?

Have been struggling with this whole day for the last 2 weeks so any help is welcome...

_Originally posted by @skalaydzhiyski in https://github.com/nautechsystems/nautilus_trader/discussions/1416_

**Comment by cjdsellers** [comment]
Hey @skalaydzhiyski 

For Interactive Brokers this should work the same way as for other instrument classes, unless CFD's aren't implemented. Maybe @benjaminsingleton knows more?

**Comment by benjaminsingleton** [comment]
@skalaydzhiyski I haven't experimented with CFD's, so let play around this week, and I'll share my findings.

**Reply by skalaydzhiyski** [reply]
Thank you so much ! I think there might be something wrong with ib_insync instead of nautilus, but I never managed to find anything for ib_insync as well when it comes to CFDs so I'm pretty much in the dark there.

**Comment by benjaminsingleton** [comment]
Unfortunately, you are correct that CFDs are not currently supported in the IB adapter in nautilus-trader. We should update the docs, so CFD is not used as an example [here](https://docs.nautilustrader.io/integrations/ib.html#instruments-contracts).

Here is a simple example script to fetch the NVIDIA CFD [contract](https://pennies.interactivebrokers.com/cstools/contract_info/v3.10/index.php?action=Conid%20Info&wlId=IB&conid=120549668&lang=en&ib_entity=) / instrument, which fails despite the TWS API returning valid data:

```
import asyncio

from nautilus_trader.adapters.interactive_brokers.common import IBContract
from nautilus_trader.adapters.interactive_brokers.historic import HistoricInteractiveBrokersClient


async def main():
    client = HistoricInteractiveBrokersClient(port=4002, client_id=1)
    await client._connect()
    await asyncio.sleep(2)

    instruments = await client.request_instruments(
        contracts=[
            IBContract(
                symbol="NVDA",
                secType="CFD",
                currency="USD",
            )
        ]
    )

    print(instruments)

if __name__ == "__main__":
    asyncio.run(main())
```

The following error is returned:

```
Traceback (most recent call last):
  File "/Users/ben/Projects/cfd-test/scripts/fetch_instrument.py", line 24, in <module>
    asyncio.run(main())
  File "/Users/ben/.pyenv/versions/3.10.13/lib/python3.10/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/Users/ben/.pyenv/versions/3.10.13/lib/python3.10/asyncio/base_events.py", line 649, in run_until_complete
    return future.result()
  File "/Users/ben/Projects/cfd-test/scripts/fetch_instrument.py", line 17, in main
    instruments = await client.request_instruments(
  File "/Users/ben/Projects/cfd-test/.venv/lib/python3.10/site-packages/nautilus_trader/adapters/interactive_brokers/historic/client.py", line 132, in request_instruments
    await provider.load_all_async()
  File "/Users/ben/Projects/cfd-test/.venv/lib/python3.10/site-packages/nautilus_trader/adapters/interactive_brokers/providers.py", line 84, in load_all_async
    await self.load_ids_async([])
  File "/Users/ben/Projects/cfd-test/.venv/lib/python3.10/site-packages/nautilus_trader/adapters/interactive_brokers/providers.py", line 104, in load_ids_async
    await self.load_async(contract)
  File "/Users/ben/Projects/cfd-test/.venv/lib/python3.10/site-packages/nautilus_trader/adapters/interactive_brokers/providers.py", line 271, in load_async
    instrument: Instrument = parse_instrument(
  File "/Users/ben/Projects/cfd-test/.venv/lib/python3.10/site-packages/nautilus_trader/adapters/interactive_brokers/parsing/instruments.py", line 145, in parse_instrument
    raise ValueError(f"Unknown {security_type=}")
ValueError: Unknown security_type='CFD'
```

TWS API returns:

<img width="680" alt="image" src="https://github.com/nautechsystems/nautilus_trader/assets/6757634/09690d57-a99e-4bf6-abb5-857e3244a13d">


So it looks like a change needs to be made here: https://github.com/nautechsystems/nautilus_trader/blob/master/nautilus_trader/adapters/interactive_brokers/parsing/instruments.py#L128

**Reply by skalaydzhiyski** [reply]
Thank you very much for the elaborate answer! Is there any chance you can show me an example of the TWS API without going through nautilus ? I didn't get it to work even just using ib_insync :/

**Reply by skalaydzhiyski** [reply]
@benjaminsingleton - btw there is no historical data for CFDs in IB using ib_insync, so the solution might be to use STK historical data for backtesting and for getting live feed when running the strategy live, but when sending orders instead of buying the underlying stock -> buy the CFD. Do you of any example of doing that following the nautilus best practices ?

thanks again
